compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -744,78 +745,92 @@ impl<'a, 'b> Context<'a, 'b> {
+        // element of original_args it's passing, and that argument's type.
+        let fmt_arg_index_and_ty = self
+            .arg_unique_types
+            .iter()
+            .enumerate()
+            .flat_map(|(i, unique_types)| unique_types.iter().map(move |ty| (i, ty)))
+            .chain(self.count_args.iter().map(|i| (*i, &Count)));
+
+        // Figure out whether there are permuted or repeated elements. If not,
+        // we can generate simpler code.
+        let nicely_ordered = fmt_arg_index_and_ty
+            .clone()
+            .is_sorted_by(|(i, _), (j, _)| (i < j).then_some(Ordering::Less));
+
+        // We want to emit:
         //
-        //      let tmp0 = &arg0;
-        //      let tmp1 = &arg1;
-        //      args_array
+        //     [ArgumentV1::new(&$arg0, …), ArgumentV1::new(&$arg1, …), …]
         //
-        // Because of #11585 the new temporary lifetime rule, the enclosing
-        // statements for these temporaries become the let's themselves.
-        // If one or more of them are RefCell's, RefCell borrow() will also
-        // end there; they don't last long enough for args_array to use them.
-        // The match expression solves the scope problem.
+        // However, it's only legal to do so if $arg0, $arg1, … were written in
+        // exactly that order by the programmer. When arguments are permuted, we
+        // want them evaluated in the order written by the programmer, not in
+        // the order provided to fmt::Arguments. When arguments are repeated, we
+        // want the expression evaluated only once.
         //
-        // Note, it may also very well be transformed to:
+        // Thus in the not nicely ordered case we emit the following instead:
         //
-        //      match arg0 {
-        //          ref tmp0 => {
-        //              match arg1 => {
-        //                  ref tmp1 => args_array } } }
+        //     match (&$arg0, &$arg1, …) {
+        //         _args => [ArgumentV1::new(_args.$i, …), ArgumentV1::new(_args.$j, …), …]
+        //     }
         //
-        // But the nested match expression is proved to perform not as well
-        // as series of let's; the first approach does.
-        let args_match = {
-            let pat = self.ecx.pat_ident(self.macsp, Ident::new(sym::_args, self.macsp));
-            let arm = self.ecx.arm(self.macsp, pat, args_array);
-            let head = self.ecx.expr(self.macsp, ast::ExprKind::Tup(heads));
-            self.ecx.expr_match(self.macsp, head, vec![arm])
-        };
+        // for the sequence of indices $i, $j, … governed by fmt_arg_index_and_ty.
+        for (arg_index, arg_ty) in fmt_arg_index_and_ty {
+            let e = &mut original_args[arg_index];
+            let span = e.span;
+            let arg = if nicely_ordered {
+                // The indices are strictly ordered so e has not been taken yet.
+                self.ecx.expr_addr_of(span, P(e.take()))
```

Comment:
This would undo the fix in #90131.

```suggestion
                self.ecx.expr_addr_of(e.span.with_ctxt(self.macsp.ctxt()), P(e.take()))
```

Reply:
Ah good catch! I'd written this prior to #90131 and only changed the span in one place during the rebase. I guess that explains why all the *"this error originates in the macro \`$crate::format_args_nl\` (in Nightly builds, run with -Z macro-backtrace for more info)"* went away. :sweat_smile: